# Loan Predictions 

The objective of this project is to build a machine learning model that can accurately predict if a borrower will pay off their loan on time or not?
The financial lending dataset was collected by a peer-to-peer lending company called Lending Club which provides an online  marketplace where borrowers match with investors in order to make credit more affordable and investing more rewarding.

##### Binary classification problem 
Select target column for ML phase, which column is best suited for this?

"The Data Science Process is about observation, model building, analysis and conclusion"
We thoroughly followed the data science process as shown below:
1. Ask questions and identify the problem
2. Data Collection
3. Data Exploration
4. Data Modeling
5. Data Analysis
6. Visualization and Presentation of Result

Data Cleaning 
Preparing Features
Making Predictions 

#### Analysis 
- key factors for predicting if a borrower will pay off their loan on time 
- 

Libraries used:
- Pandas
- Numpy
- Scikit-learn 

### Summary of Results
- randomforest classifier 
- logistic regression 




### Data Import

In [57]:
#import necessary libraries 
import pandas as pd
import numpy as np


# import libraries to perform logistic regression 
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import cross_val_predict

# import libraries to perform randomforestclassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.cross_validation import cross_val_predict

loans= pd.read_csv("loans_2007.csv")
loans.drop_duplicates()

C:\Users\soohy\AppData\Local\Enthought\Canopy\edm\envs\User\lib\site-packages\IPython\core\interactiveshell.py:2717: DtypeWarning: Columns (0) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


id  member_id  loan_amnt  \
0                                              1077501  1296599.0     5000.0   
1                                              1077430  1314167.0     2500.0   
2                                              1077175  1313524.0     2400.0   
3                                              1076863  1277178.0    10000.0   
4                                              1075358  1311748.0     3000.0   
5                                              1075269  1311441.0     5000.0   
6                                              1069639  1304742.0     7000.0   
7                                              1072053  1288686.0     3000.0   
8                                              1071795  1306957.0     5600.0   
9                                              1071570  1306721.0     5375.0   
10                                             1070078  1305201.0     6500.0   
11                                             1069908  1305008.0    12000.0   
12                                             1064687  1298717.0     9000.0   
13                                             1069866  1304956.0     3000.0   
14                                             1069057  1303503.0    10000.0   
15                                             1069759  1304871.0     1000.0   
16                                             1065775  1299699.0    10000.0   
17                                             1069971  1304884.0     3600.0   
18                                             1062474  1294539.0     6000.0   
19                                             1069742  1304855.0     9200.0   
20                                             1069740  1284848.0    20250.0   
21                                             1039153  1269083.0    21000.0   
22                                             1069710  1304821.0    10000.0   
23                                             1069700  1304810.0    10000.0   
24                                             1069559  1304634.0     6000.0   
25                                             1069697  1273773.0    15000.0   
26                                             1069800  1304679.0    15000.0   
27                                             1069657  1304764.0     5000.0   
28                                             1069799  1304678.0     4000.0   
29                                             1047704  1278806.0     8500.0   
...                                                ...        ...        ...   
42508                                            91175    91170.0     6000.0   
42509                                            91126    91067.0     5350.0   
42510                                            91023    70879.0     1900.0   
42511                                            90106    90090.0    10000.0   
42512                                            89258    80039.0     2000.0   
42513                                            88637    88629.0     6000.0   
42514                                            88046    88023.0     4400.0   
42515                                            85961    85923.0     1200.0   
42516                                            85818    85802.0     5000.0   
42517                                            85781    85727.0     1400.0   
42518                                            85675    85667.0     1000.0   
42519                                            84670    79576.0     5000.0   
42520                                            84098    84091.0     2500.0   
42521                                            83979    83974.0     3000.0   
42522                                            83489    83471.0     2600.0   
42523                                            83185    83132.0     1000.0   
42524                                            74014    73890.0     6450.0   
42525                                            76629    76623.0     1275.0   
42526                                            810

In [58]:
print(loans.iloc[0])

id                                1077501
member_id                      1.2966e+06
loan_amnt                            5000
funded_amnt                          5000
funded_amnt_inv                      4975
term                            36 months
int_rate                           10.65%
installment                        162.87
grade                                   B
sub_grade                              B2
emp_title                             NaN
emp_length                      10+ years
home_ownership                       RENT
annual_inc                          24000
verification_status              Verified
issue_d                          Dec-2011
loan_status                    Fully Paid
pymnt_plan                              n
purpose                       credit_card
title                            Computer
zip_code                            860xx
addr_state                             AZ
dti                                 27.65
delinq_2yrs                       

In [59]:
print(len(loans.columns))

52


### Data Cleaning 

Break up the dataset into three sets to focus on the columns in order to better understand the data and eliminate any unneccesaary or irrelevant columns. Look at the data dictionary provided. 


In [60]:
# Subset the data using an .iloc indexer
loans1 = loans.iloc[:,0:17]
loans2= loans.iloc[:,17:35]
loans3= loans.iloc[:,35:52] 

##### 1st Set (Columns 0-18)

In [61]:
loans1.head()

id  member_id  loan_amnt  funded_amnt  funded_amnt_inv        term  \
0  1077501  1296599.0     5000.0       5000.0           4975.0   36 months   
1  1077430  1314167.0     2500.0       2500.0           2500.0   60 months   
2  1077175  1313524.0     2400.0       2400.0           2400.0   36 months   
3  1076863  1277178.0    10000.0      10000.0          10000.0   36 months   
4  1075358  1311748.0     3000.0       3000.0           3000.0   60 months   

  int_rate  installment grade sub_grade                 emp_title emp_length  \
0   10.65%       162.87     B        B2                       NaN  10+ years   
1   15.27%        59.83     C        C4                     Ryder   < 1 year   
2   15.96%        84.33     C        C5                       NaN  10+ years   
3   13.49%       339.31     C        C1       AIR RESOURCES BOARD  10+ years   
4   12.69%        67.79     B        B5  University Medical Group     1 year   

  home_ownership  annual_inc verification_status   issue_d  loan_status  
0           RENT     24000.0            Verified  Dec-2011   Fully Paid  
1           RENT     30000.0     Source Verified  Dec-2011  Charged Off  
2           RENT     12252.0        Not Verified  Dec-2011   Fully Paid  
3           RENT     49200.0     Source Verified  Dec-2011   Fully Paid  
4           RENT     80000.0     Source Verified  Dec-2011      Current

After analyzing each column, we can conclude that the following features need to be removed:

- **id**: randomly generated field by Lending Club for unique identification purposes only
- **member_id**: also a randomly generated field by Lending Club for unique identification purposes only
- **funded_amnt**: leaks data from the future (after the loan is already started to be funded)
- **funded_amnt_inv**: also leaks data from the future (after the loan is already started to be funded)
- **grade**: contains redundant information as the interest rate column (int_rate)
- **sub_grade**: also contains redundant information as the interest rate column (int_rate)
- **emp_title**: requires other data and a lot of processing to potentially be useful
- **issue_d**: leaks data from the future (after the loan is already completed funded)

Recall that Lending Club assigns a grade and a sub-grade based on the borrower's interest rate. While the grade and sub_grade values are categorical, the int_rate column contains continuous values, which are better suited for machine learning.

In [62]:
# drop the unneccessary columns 
loans=loans.drop( ["id", "member_id", "funded_amnt", "funded_amnt_inv", "grade", "sub_grade", "emp_title", "issue_d"], axis=1)


##### 2nd Set (Columns 18-36)

In [63]:
loans2.head()

pymnt_plan         purpose                 title zip_code addr_state    dti  \
0          n     credit_card              Computer    860xx         AZ  27.65   
1          n             car                  bike    309xx         GA   1.00   
2          n  small_business  real estate business    606xx         IL   8.72   
3          n           other              personel    917xx         CA  20.00   
4          n           other              Personal    972xx         OR  17.94   

   delinq_2yrs earliest_cr_line  inq_last_6mths  open_acc  pub_rec  revol_bal  \
0          0.0         Jan-1985             1.0       3.0      0.0    13648.0   
1          0.0         Apr-1999             5.0       3.0      0.0     1687.0   
2          0.0         Nov-2001             2.0       2.0      0.0     2956.0   
3          0.0         Feb-1996             1.0      10.0      0.0     5598.0   
4          0.0         Jan-1996             0.0      15.0      0.0    27783.0   

  revol_util  total_acc initial_list_status  out_prncp  out_prncp_inv  \
0      83.7%        9.0                   f       0.00           0.00   
1       9.4%        4.0                   f       0.00           0.00   
2      98.5%       10.0                   f       0.00           0.00   
3        21%       37.0                   f       0.00           0.00   
4      53.9%       38.0                   f     461.73         461.73   

    total_pymnt  
0   5863.155187  
1   1008.710000  
2   3005.666844  
3  12231.890000  
4   3581.120000

Within this group of columns, we need to drop the following columns:

- **zip_code**: redundant with the addr_state column since only the first 3 digits of the 5 digit zip code are visible (which only can be used to identify the state the borrower lives in)
- **out_prncp**: leaks data from the future, (after the loan already started to be paid off)
- **out_prncp_inv**: also leaks data from the future, (after the loan already started to be paid off)
- **total_pymnt**: also leaks data from the future, (after the loan already started to be paid off)

The out_prncp and out_prncp_inv both describe the outstanding principal amount for a loan, which is the remaining amount the borrower still owes. These 2 columns as well as the total_pymnt column describe properties of the loan after it's fully funded and started to be paid off. This information isn't available to an investor before the loan is fully funded and we don't want to include it in our model.


In [64]:
# drop the unneccessary columns 
loans = loans.drop(["zip_code", "out_prncp", "out_prncp_inv", "total_pymnt"], axis=1)


##### 3rd Set (Columns 18-52)

In [65]:
loans3.head()

total_pymnt_inv  total_rec_prncp  total_rec_int  total_rec_late_fee  \
0          5833.84          5000.00         863.16                0.00   
1          1008.71           456.46         435.17                0.00   
2          3005.67          2400.00         605.67                0.00   
3         12231.89         10000.00        2214.92               16.97   
4          3581.12          2538.27        1042.85                0.00   

   recoveries  collection_recovery_fee last_pymnt_d  last_pymnt_amnt  \
0        0.00                     0.00     Jan-2015           171.62   
1      117.08                     1.11     Apr-2013           119.66   
2        0.00                     0.00     Jun-2014           649.91   
3        0.00                     0.00     Jan-2015           357.48   
4        0.00                     0.00     Jun-2016            67.79   

  last_credit_pull_d  collections_12_mths_ex_med  policy_code  \
0           Jun-2016                         0.0          1.0   
1           Sep-2013                         0.0          1.0   
2           Jun-2016                         0.0          1.0   
3           Apr-2016                         0.0          1.0   
4           Jun-2016                         0.0          1.0   

  application_type  acc_now_delinq  chargeoff_within_12_mths  delinq_amnt  \
0       INDIVIDUAL             0.0                       0.0          0.0   
1       INDIVIDUAL             0.0                       0.0          0.0   
2       INDIVIDUAL             0.0                       0.0          0.0   
3       INDIVIDUAL             0.0                       0.0          0.0   
4       INDIVIDUAL             0.0                       0.0          0.0   

   pub_rec_bankruptcies  tax_liens  
0                   0.0        0.0  
1                   0.0        0.0  
2                   0.0        0.0  
3                   0.0        0.0  
4                   0.0        0.0

 drop the following columns:
- **total_pymnt_inv**: also leaks data from the future, (after the loan already started to be paid off)
- **total_rec_prncp**: also leaks data from the future, (after the loan already started to be paid off)
- **total_rec_int**: leaks data from the future, (after the loan already started to be paid off),
- **total_rec_late_fee**: also leaks data from the future, (after the loan already started to be paid off),
- **recoveries**: also leaks data from the future, (after the loan already started to be paid off),
- **collection_recovery_fee**: also leaks data from the future, (after the loan already started to be paid off),
- **last_pymnt_d**: also leaks data from the future, (after the loan already started to be paid off),
- **last_pymnt_amnt**: also leaks data from the future, (after the loan already started to be paid off).

All of these columns leak data from the future, meaning that they're describing aspects of the loan after it's already been fully funded and started to be paid off by the borrower.


In [66]:
# drop the unneccessary columns 
loans = loans.drop(["total_pymnt_inv", "total_rec_prncp", "total_rec_int", "total_rec_late_fee", "recoveries", "collection_recovery_fee", "last_pymnt_d", "last_pymnt_amnt"], axis=1)


In [67]:
print(loans.iloc[0])
print(len(loans.columns))

loan_amnt                            5000
term                            36 months
int_rate                           10.65%
installment                        162.87
emp_length                      10+ years
home_ownership                       RENT
annual_inc                          24000
verification_status              Verified
loan_status                    Fully Paid
pymnt_plan                              n
purpose                       credit_card
title                            Computer
addr_state                             AZ
dti                                 27.65
delinq_2yrs                             0
earliest_cr_line                 Jan-1985
inq_last_6mths                          1
open_acc                                3
pub_rec                                 0
revol_bal                           13648
revol_util                          83.7%
total_acc                               9
initial_list_status                     f
last_credit_pull_d               J

### Target column 

We now need to decide on a target column that we want to use for modeling.

We should use the loan_status column, since it's the only column that directly describes if a loan was paid off on time, had delayed payments, or was defaulted on the borrower. Currently, this column contains text values and we need to convert it to a numerical one for training a model. Let's explore the different values in this column and come up with a strategy for converting the values in this column.

In [68]:
# Explore loan_status unique values
print(loans["loan_status"].value_counts)



<bound method IndexOpsMixin.value_counts of 0                                               Fully Paid
1                                              Charged Off
2                                               Fully Paid
3                                               Fully Paid
4                                                  Current
5                                               Fully Paid
6                                               Fully Paid
7                                               Fully Paid
8                                              Charged Off
9                                              Charged Off
10                                              Fully Paid
11                                              Fully Paid
12                                             Charged Off
13                                              Fully Paid
14                                             Charged Off
15                                              Fully Paid
16          

There are 8 different possible values for the loan_status column. You can read about most of the different loan statuses on the Lending Clube webste. The 2 values that start with "Does not meet the credit policy" aren't explained unfortunately. .

We've compiled the explanation for each column as well as the counts in the Dataframe in the following table:

From the investor's perspective, we're interested in trying to predict which loans will be paid off on time and which ones won't be. Only the Fully Paid and Charged Off values describe the final outcome of the loan. The other values describe loans that are still on going and where the jury is still out on if the borrower will pay back the loan on time or not. While the Default status resembles the Charged Off status, in Lending Club's eyes, loans that are charged off have essentially no chance of being repaid while default ones have a small chance. 

Since we're interested in being able to predict which of these 2 values a loan will fall under, we can treat the problem as a binary classification one. Let's remove all the loans that don't contain either Fully Paid and Charged Off as the loan's status and then transform the Fully Paid values to 1 for the positive case and the Charged Off values to 0 for the negative case. While there are a few different ways to transform all of the values in a column, we'll use the Dataframe method replace. According to the documentation, we can pass the replace method a nested mapping dictionary in the following format:

Lastly, one thing we need to keep in mind is the class imbalance between the positive and negative cases. While there are 33,136 loans that have been fully paid off, there are only 5,634 that were charged off. This class imbalance is a common problem in binary classification and during training, the model ends up having a strong bias towards predicting the class with more observations in the training set and will rarely predict the class with less observations. The stronger the imbalance, the more biased the model becomes. There are a few different ways to tackle this class imbalance, which we'll explore later.

In [69]:
# Remove all rows from loans that contain values other than Fully Paid or Charged Off for the loan_status column.
a = ["Fully Paid", "Charged Off"]
loans = loans[loans["loan_status"].isin(a)]

# Use the Dataframe method replace to replace:Fully Paid with 1 and Charged Off with 0
mapping_dict = {
   "Fully Paid" : 1,
   "Charged Off" : 0
}
loans = loans.replace(mapping_dict)


### Feature Columns


For each column:
Use the Series method dropna to remove any null values and then use the Series method unique to return the set of non-null unique values
Use the len() function to return the number of values in that set
Append the column to drop_columns if it contains only 1 unique value



In [75]:
# Remove any columns from loans that contain only one unique value:

#Create an empty list, drop_columns to keep track of which columns you want to drop
orgi_columns = loans
drop_columns = []

# For each column: Use the Series method dropna to remove any null values and then use the Series method unique to return the set of non-null unique values
# Use the len() function to return the number of values in that set
# Append the column to drop_columns if it contains only 1 unique value

for column in loans:
    non_null = loans[column].dropna()
    unique_non_null = non_null.unique()
    num_true_unique = len(unique_non_null)
    if num_true_unique == 1:
        drop_columns.append(column)
# Use the Dataframe method drop to remove the columns in drop_columns from loans     
loans = loans.drop( drop_columns, axis =1)

# Use the print function to display drop_columns so we know which ones were removed
print(drop_columns)


['pymnt_plan', 'initial_list_status', 'collections_12_mths_ex_med', 'policy_code', 'application_type', 'acc_now_delinq', 'chargeoff_within_12_mths', 'delinq_amnt', 'tax_liens']


It looks we we were able to remove 9 more columns since they only contained 1 unique value. So far,  we started to become familiar with the columns in the dataset and removed many columns that aren't useful for modeling. We also selected our target column and decided to focus our modeling efforts on binary classification. In the next mission, we'll explore the individual features in greater depth and work towards training our first machine learning model.


In this mission, we performed the last amount of data preparation necessary to start training machine learning models. We converted all of the columns to numerical values because those are the only type of value scikit-learn can work with. In the next mission, we'll experiment with training models and evaluating accuracy using cross-validation.


 In this mission, we'll prepare the data for machine learning by focusing on handling missing values, converting categorical columns to numeric columns, and removing any other extraneous columns we encounter throughout this process.

This is because the mathematics underlying most machine learning models assumes that the data is numerical and contains no missing values. To reinforce this requirement, scikit-learn will return an error if you try to train a model using data that contain missing values or non-numeric values when working with models like linear regression and logistic regression.

Let's start by computing the number of missing values and come up with a strategy for handling them. Then, we'll focus on the categorical columns.


In [76]:
# return the number of missing values across the Dataframe
# use the Pandas Dataframe method isnull to return a Dataframe containing Boolean values:
null_counts = loans.isnull().sum()
# use the Pandas Dataframe method sum to calculate the number of null values in each column.
print(null_counts)

loan_amnt                 0
term                      0
int_rate                  0
installment               0
emp_length                0
home_ownership            0
annual_inc                0
verification_status       0
loan_status               0
purpose                   0
title                    10
addr_state                0
dti                       0
delinq_2yrs               0
earliest_cr_line          0
inq_last_6mths            0
open_acc                  0
pub_rec                   0
revol_bal                 0
revol_util               50
total_acc                 0
last_credit_pull_d        2
pub_rec_bankruptcies    697
dtype: int64


While most of the columns have 0 missing values, 2 columns have 50 or less rows with missing values, and 1 column, pub_rec_bankruptcies, contains 697 rows with missing values. Let's remove columns entirely where more than 1% of the rows for that column contain a null value. In addition, we'll remove the remaining rows containing null values.



This means that we'll keep the following columns and just remove rows containing missing values for them:
title
revol_util
last_credit_pull_d
and drop the pub_rec_bankruptcies column entirely since more than 1% of the rows have a missing value for this column.

Let's use the strategy of removing the pub_rec_bankruptcies column first then removing all rows containing any missing values at all to cover both of these cases. This way, we only remove the rows containing missing values for the title and revol_util columns but not the pub_rec_bankruptcies column.


In [78]:
# Drop the pub_rec_bankruptcies column entirely since more than 1% of the rows have a missing value for this column
loans = loans.drop(["pub_rec_bankruptcies"], axis=1)
# Remove all rows from loans containing any missing values
loans = loans.dropna()
# Use the dtypes attribute followed by the value_counts() method to return the counts for each column data type.
print(loans.dtypes.value_counts())


object     11
float64    10
int64       1
dtype: int64


While the numerical columns can be used natively with scikit-learn, the object columns that contain text need to be converted to numerical data types. Let's return a new Dataframe containing just the object columns so we can explore them in more depth

In [79]:
# Use the Dataframe method select_dtypes to select only the columns of object type from loans and assign the resulting Dataframe object_columns_df.
object_columns_df = loans.select_dtypes(include=['object'])
# display a sample row to get a better sense of how the values in each column are formatted.
print(object_columns_df.iloc[0])


term                     36 months
int_rate                    10.65%
emp_length               10+ years
home_ownership                RENT
verification_status       Verified
purpose                credit_card
title                     Computer
addr_state                      AZ
earliest_cr_line          Jan-1985
revol_util                   83.7%
last_credit_pull_d        Jun-2016
Name: 0, dtype: object


Some of the columns seem like they represent categorical values, but we should confirm by checking the number of unique values in those columns:

- **home_ownership**: home ownership status, can only be 1 of 4 categorical values according to the data dictionary,
- **verification_status**: indicates if income was verified by Lending Club,
- **emp_length**: number of years the borrower was employed upon time of application,
- **term**: number of payments on the loan, either 36 or 60,
- **addr_state**: borrower's state of residence,
- **purpose**: a category provided by the borrower for the loan request,
- **title**: loan title provided the borrower,

There are also some columns that represent numeric values, that need to be converted:

- **int_rate**: interest rate of the loan in %,
- **revol_util**: revolving line utilization rate or the amount of credit the borrower is using relative to all available credit, read more here.
Based on the first row's values for purpose and title, it seems like these columns could reflect the same information. Let's explore the unique value counts separately to confirm if this is true.

Lastly, some of the columns contain date values that would require a good amount of feature engineering for them to be potentially useful:

- **earliest_cr_line**: The month the borrower's earliest reported credit line was opened,
- **last_credit_pull_d**: The most recent month Lending Club pulled credit for this loan.

Since these date features require some feature engineering for modeling purposes, let's remove these date columns from the Dataframe.

In [ ]:
# Display the unique value counts for the following columns
cols = ['home_ownership', 'verification_status', 'emp_length', 'term', 'addr_state']   
for item in cols:
    print(loans[item].value_counts())


The home_ownership, verification_status, emp_length, term, and addr_state columns all contain multiple discrete values. We should clean the emp_length column and treat it as a numerical one since the values have ordering (2 years of employment is less than 8 years).

First, let's look at the unique value counts for the purpose and title columns to understand which column we want to keep.



In [ ]:
# Use the value_counts method and the print function to display the unique values in the following columns:
print(loans["purpose"].value_counts())
print(loans["title"].value_counts())

The home_ownership, verification_status, emp_length, and term columns each contain a few discrete categorical values. We should encode these columns as dummy variables and keep them.

It seems like the purpose and title columns do contain overlapping information but we'll keep the purpose column since it contains a few discrete values. In addition, the title column has data quality issues since many of the values are repeated with slight modifications (e.g. Debt Consolidation and Debt Consolidation Loan and debt consolidation).

We can use the following mapping to clean the emp_length column:

"10+ years": 10
"9 years": 9
"8 years": 8
"7 years": 7
"6 years": 6
"5 years": 5
"4 years": 4
"3 years": 3
"2 years": 2
"1 year": 1
"< 1 year": 0
"n/a": 0
We erred on the side of being conservative with the 10+ years, < 1 year and n/a mappings. We assume that people who may have been working more than 10 years have only really worked for 10 years. We also assume that people who've worked less than a year or if the information is not available that they've worked for 0. This is a general heuristic but it's not perfect.

Lastly, the addr_state column contains many discrete values and we'd need to add 49 dummy variable columns to use it for classification. This would make our Dataframe much larger and could slow down how quickly the code runs. Let's remove this column from consideration.

In [ ]:
# Remove the last_credit_pull_d, addr_state, title, and earliest_cr_line columns from loans.
loans = loans.drop(columns = ["last_credit_pull_d", "addr_state", "title", "earliest_cr_line"])

# use the following mapping to clean the emp_length column
mapping_dict = {
    "emp_length": {
        "10+ years": 10,
        "9 years": 9,
        "8 years": 8,
        "7 years": 7,
        "6 years": 6,
        "5 years": 5,
        "4 years": 4,
        "3 years": 3,
        "2 years": 2,
        "1 year": 1,
        "< 1 year": 0,
        "n/a": 0
    }
}
# Use the replace method to clean the emp_length column
loans = loans.replace(mapping_dict)


# Convert the int_rate and revol_util columns to float columns by: 
# Using the str acessor followed by the rstrip string method to strip the right trailing percent sign (%):
# use the astype method to convert to the float type.

loans['int_rate'] = loans['int_rate'].str.rstrip('%').astype(float)
loans['revol_util'] = loans['revol_util'].str.rstrip('%').astype(float)




Let's now encode the home_ownership, verification_status, purpose, and term columns as dummy variables so we can use them in our model. We first need to use the Pandas get_dummies method to return a new Dataframe containing a new column for each dummy variable:

### Convert Categorical variables to Numerical 

In [81]:
# encode the following categorical columns into dummy variables to use them in the model
cols = ["home_ownership", "verification_status", "purpose", "term"]
# Use the get_dummies function to return a Dataframe containing the dummy columns.
dummy_df = pd.get_dummies(loans[cols])
# We can then use the concat method to add these dummy columns back to the original Dataframe:
loans = pd.concat([loans, dummy_df], axis=1)
# drop the original columns entirely using the drop method:
loans = loans.drop( cols, axis=1)

In this mission, we performed the last amount of data preparation necessary to start training machine learning models. We converted all of the columns to numerical values because those are the only type of value scikit-learn can work with. In the next mission, we'll experiment with training models and evaluating accuracy using cross-validation.

# Exploratory Data Analysis

In [ ]:
loans.head()
loans.describe()

Correlations & Other Features
Having validated and invalidated some of our hypotheses in the initial exploration, we've realized that a key factor is understanding which features correlate largely with the target variable.

This can be understood numerically, with the cross-correlation matrix! Let's take a look:



In [ ]:
loans.corr()

# Conclusion